pbxd/pbx/definity.py

```diff
@@ -0,0 +1,411 @@
+            raise Exception('Timeout on command prompt verify:\n{}'.format(self.session.before))
+        elif index == 1:  # EOF
+            self.logger.error('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
+            raise Exception('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
+
+        self.connected_termtype = termtype
+
+    def _ossi_object(self, response_fields, response_data):
+        """
+        Convert the OSSI field and data lists to a dictionary.
+        """
+        ossi_obj = {}
+        for i, field in enumerate(response_fields):
+            ossi_obj[field] = response_data[i]
+        if len(response_fields) != len(ossi_obj):
+            # there have been cases of duplicate field ids in some commands
+            self.logger.error('duplicate field ids detected {} != {}'.format(response_fields, ossi_obj.keys()))
+        return ossi_obj
+
+    def ossi_command(self, command, fields=None, debug=False):
+        """
+        Send a command to the PBX and return the result.
+
+        Provide a dictionary of field ids and values to change fields or return
+        specific fields.
+
+        One way to identify field ids is to run a "display" command and then
+        compare it to the output of the same command in a VT220 terminal.
+
+        Note: there have been cases of duplicate field ids from the PBX so the
+        data_list is available without ids if needed.
+        """
+        # switch back to the original ossi OSSI terminal type
+        self._select_termtype('ossi4')
+
+        # Send OSSI command
+        self.logger.info('command: {}'.format(command))
+        self.session.sendline('c {}'.format(command))  # command
+
+        # if no fields are specificed the pbx returns all fields
+        if fields is not None and len(fields) > 0:
+            self.logger.debug('fields: {}'.format(fields))
+            ids = ('\t'.join(sorted(fields))).strip()
+            data = ('\t'.join([fields[k] for k in sorted(fields)])).strip()
+            self.logger.debug('send: f{}'.format(ids))
+            self.session.sendline('f{}'.format(ids))  # fields
+            self.logger.debug('send: d{}'.format(data))
+            self.session.sendline('d{}'.format(data))  # data
+
+        self.session.sendline('t')  # command terminator
+
+        # Read the response and collect the OSSI objects
+        response_fields = []
+        response_data = []
+        response_error = None
+        complete_output = False
+        ossi_objects = []
+        raw_lines = []
+        while not complete_output:
+            index = self.session.expect(
+                [
+                    pexpect.TIMEOUT,
+                    pexpect.EOF,
+                    r'f[\S\t]+[\r\n]+',
+                    r'd[\S\t ]*[\r\n]+',
+                    r'e[\S\t ]+[\r\n]+',
```

Comment:
These regexes seem weird to me and I don't know why they're all different:

`[\S\t]+`: one or more non-space characters or tabs
`[\S\t ]*`: zero or more non-space or tabs or spaces
`[\S\t ]+` one or more non-space characters or tabs or spaces

etc.

In any case I don't think checking for six different formats in the `expect` line is necessary, especially since you split the lines up again below.

I would probably just do something like `[cdefnt].*\r\n` as the main pattern to be matched and then handle the various line types in a different way.

Reply:
Unfortunately they are all slighty different and need to be distinct to match correctly.

Reply:
I had a further comment here but then I clicked the wrong button at the wrong time and all my work went away…here's the short version

A nicer factoring for some of this would be a pure `parse_ossi_response_line()` function that takes a string and returns a parsed structure. It'd be a lot easier to plow a bunch of test cases drawn from real traffic into that.

Reply:
I'm not sure how this would work as the parsing is pretty simple and it's the unique action for each type of line that matters. For now I'm happy with it this way, but if there's a simpler way we could look at that in the future.